modules/reindex/src/main/java/org/elasticsearch/reindex/ReindexValidator.java

```diff
@@ -138,12 +139,32 @@ static void validateAgainstAliases(
              */
             target = indexNameExpressionResolver.concreteWriteIndex(clusterState, destination).getName();
         }
-        for (String sourceIndex : indexNameExpressionResolver.concreteIndexNames(clusterState, source)) {
+        SearchRequest filteredSource = skipRemoteIndexNames(source);
+        if (filteredSource.indices().length == 0) {
+            return;
+        }
+        String[] sourceIndexNames = indexNameExpressionResolver.concreteIndexNames(clusterState, filteredSource);
+        for (String sourceIndex : sourceIndexNames) {
             if (sourceIndex.equals(target)) {
                 ActionRequestValidationException e = new ActionRequestValidationException();
                 e.addValidationError("reindex cannot write into an index its reading from [" + target + ']');
                 throw e;
             }
         }
     }
+
+    private static SearchRequest skipRemoteIndexNames(SearchRequest source) {
+        return new SearchRequest(source).indices(
+            Arrays.stream(source.indices()).filter(name -> isRemoteExpression(name) == false).toArray(String[]::new)
+        );
+    }
+
+    private static boolean isRemoteExpression(String expression) {
+        // remote index name contains ':' symbol inside their names (remote cluster name separator), e.g. cluster0:index-name
+        // in the same time date-math `expression` can also contain ':' symbol inside its name
+        // to distinguish between those two, given `expression` is pre-evaluated using date-math resolver
+        // after evaluation date-math `expression` should not contain ':' symbol
+        // otherwise if `expression` is legit remote name, ':' symbol remains
+        return IndexNameExpressionResolver.resolveDateMathExpression(expression).contains(":");
```

Comment:
Reindex API does piggyback Search API which should support complex date-math index names ([doc](https://www.elastic.co/guide/en/elasticsearch/reference/current/api-conventions.html#api-date-math-index-names)) but there is a bug there. Reindex API inherits the bug.

Since to decide if given index name is remote we lookup for semicolon, this code has to eval name in advance.